tests/integration/application-charm/src/charm.py

```diff
@@ -0,0 +1,139 @@
+#!/usr/bin/env python3
+# Copyright 2022 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Application charm that connects to the mysqlrouter charms.
+
+This charm is meant to be used only for testing
+the mysqlrouter requires-provides relation.
+"""
+
+import itertools
+import json
+import logging
+
+from charms.data_platform_libs.v0.database_requires import (
+    DatabaseCreatedEvent,
+    DatabaseRequires,
+)
+from charms.operator_libs_linux.v1 import systemd
+from connector import MySQLConnector
+from ops.charm import ActionEvent, CharmBase, StartEvent
+from ops.main import main
+from ops.model import ActiveStatus, WaitingStatus
+
+from utils import generate_random_chars
+
+logger = logging.getLogger(__name__)
+
+PEER = "application-peers"
+REMOTE = "database"
+MYSQL_ROUTER_SERVICE_NAME = "mysqlrouter.service"
+
+
+class ApplicationCharm(CharmBase):
+    """Application charm that relates to MySQL Router charm."""
+
+    def __init__(self, *args):
+        super().__init__(*args)
+
+        self.framework.observe(self.on.start, self._on_start)
+
+        self.database_name = f"{self.app.name.replace('-', '_')}_test_database"
+        self.database_requires = DatabaseRequires(self, REMOTE, self.database_name)
+        self.framework.observe(
+            self.database_requires.on.database_created, self._on_database_created
+        )
+        self.framework.observe(self.on.get_inserted_data_action, self._get_inserted_data)
+
+    # =======================
+    #  Handlers
+    # =======================
+
+    def _on_start(self, event: StartEvent) -> None:
+        """Handle the start event by setting the charm in waiting status."""
+        self.unit.status = WaitingStatus("Waiting for mysqlrouter relation")
+
+        peer_databag = self._peers.data[self.app]
+        database_config = peer_databag.get("database_config")
+        inserted_value = peer_databag.get("inserted_value")
+
+        if database_config and inserted_value:
+            database_config = json.loads(database_config)
+
+            if not systemd.service_running(MYSQL_ROUTER_SERVICE_NAME):
```

Comment:
This is an indicator that the mysqlrouter service has been bootstrapped and started in the scaling unit (since the user/password credentials are already in the relation databag). Essentially, checking that mysqlrouter is running prevents race conditions where the application unit is provisioned and is executed before the subordinate mysqlrouter unit (where the bootstrap happens)

Reply:
Hmm, interesting. Thanks for sharing that info.